packages/flutter/lib/src/cupertino/text_selection.dart

```diff
@@ -305,6 +431,12 @@ class _CupertinoTextSelectionControls extends TextSelectionControls {
     assert(type != null);
     return null;
   }
+
+  void _updateTriangleOffsetX(double triangleOffsetX) {
+    if (_toolbarStateGlobalKey.currentState != null) {
+      _toolbarStateGlobalKey.currentState.updateTriangleOffsetX(triangleOffsetX);
```

Comment:
This seems to me like it should just be passed into _TextSelectionToolbar like a normal parameter rather than manually updated like this.  Maybe you were running into problems that caused you to use this pattern and a timer?

Reply:
The exact position of the triangle requires the true width of the toolbar to be determined to prevent the triangle from appearing outside the toolbar.

Within the _TextSelectionToolbar class, we only know the number and type of menus to display but don't know the true width of the toolbar. Only the `childSize` parameter in the `Offset getPositionForChild(Size size, Size childSize)` method in `_TextSelectionToolbarLayout` class is the final true width of the toolbar, so it is necessary in `_TextSelectionToolbarLayout` class to calculate the true position of the triangle and pass the data out. 

As for why a Timer is needed to delay execution of the setState method by 50 milliseconds? Immediate execution causes the toolbar to always appear in the upper left corner of the screen.

Of course, the current implementation is not a very good solution, I hope we can discuss a better one.

Reply:
@jiisd I just dug into this and I think I agree with @LongCatIsLooong that [MultiChildLayoutDelegate](https://docs.flutter.io/flutter/rendering/MultiChildLayoutDelegate-class.html) is the way to go.  You should be able to position the triangle inside the delegate without having to pass around triangleOffsetX and without having to use a timer.

That will require a bit of refactoring existing code.  Let me know if you don't mind taking that on, otherwise I can take over the PR and do it myself if you want.

Reply:
Although this solution now achieves this effect, it is very difficult to understand, and the responsibilities of the class are not clear.;(

I think the MultiChildLayoutDelegate approach is better than the current one. Of course, it would be better if you had time to work on it together.🤝

Reply:
Hi @jiisd sorry for the late response. That sounds good to me! I'm going to work on this and create a pull request based on your changes. I'll let you know when the PR is open.